### DISCUSSION TITLE
Exposing Client Secrets in Network Calls

### DISCUSSION THREAD

**Discussion by IsanjaleeInivos** [vercel/next.js#63958] [Help]
### Summary

**Hello Next.js Community, 👋🏻**

I've encountered an issue related to client secret exposure in network calls while using Next.js framework. I have attempted various solutions but have been unable to find a resolution. Below, I provide a detailed description of the problem along with the relevant code snippet:

When utilizing the useLoadScript hook from @react-google-maps/api within a client-side component in Next.js, the Google Maps API key (NEXT_PUBLIC_GOOGLE_API_KEY) is exposed in network calls. This presents a security risk as it exposes the client secret in the browser's network tab.

I am seeking guidance on how to prevent the exposure of client secrets in network calls from Next.js applications. Ideally, I would like to find a solution that ensures the security of sensitive information such as API keys.

Any assistance or advice on resolving this issue would be greatly appreciated.

### Additional information

```js
Code Snippet:

'use client';
import { useLoadScript } from '@react-google-maps/api';

export function TripLayoutClient({ children }: { children: React.ReactNode }) {
  const { isLoaded, loadError } = useLoadScript({
    googleMapsApiKey: process.env.NEXT_PUBLIC_GOOGLE_API_KEY!,
    libraries: ['places'],
  });

  if (loadError) return <div>Map cannot be loaded right now, sorry.</div>;
  if (!isLoaded) return <div>Loading Maps...</div>;

  return <>{children}</>;
}

In the network tab of the browser's developer tools, the API call to Google Maps API includes the exposed API key:

https://maps.googleapis.com/maps/api/js?key={YOUR_API_KEY}&v=weekly&libraries=places&callback=initMap
```

**Comment by icyJoseph** [comment]
Could you rephrase your issue? IIRC the library you are using constructs a `URL`, with query params, one of which is the maps API key, which reaches out to the maps API.

- https://developers.google.com/maps/api-security-best-practices#restrict_apikey

I guess, you could proxy to the Maps API through a Next.js endpoint, which would add the API key to the query params, and give it back to you, but that's easier said than done. You'd need to use this library, https://github.com/googlemaps/js-api-loader/blob/main/src/index.ts#L285, instead, and work your way into make it work with your page.

**Comment by theoephraim** [comment]
Since google maps is often loaded on the client side and much of the interactions happen on the client only, I believe they intend for these api keys to be somewhat public - however this comes with the caveat that you must set up the correct restrictions for others to not use it.

Some more info available here - https://ozguralp.medium.com/unauthorized-google-maps-api-key-usage-cases-and-why-you-need-to-care-1ccb28bf21e
